File: 0227-Basic Calculator II/0227-Basic Calculator II.cpp
class Solution {
public:
    int calculate(string s) {
        int n = s.size();
        stack<int> St;
        int num = 0;
        char preop = '+';
        for (int i = 0; i < n; ++i) {
            char c = s[i];
            if (isdigit(c)) {
                num = num * 10 + (c - '0');
            }
            if (i == n - 1 || (!isdigit(c) && c != ' ')) {
                if (preop == '+') {
                    St.push(num);
                }
                else if (preop == '-') {
                    St.push(-num);
                }
                else if (preop == '*') {
                    int temp = St.top();
                    St.pop();
                    St.push(temp * num);
                }
                else {
                    int temp = St.top();
                    St.pop();
                    St.push(temp / num);
                }
                
                preop = c;
                num = 0;
            }
        }
        
        int result = 0;
        while (!St.empty()) {
            result += St.top();
            St.pop();
        }
        
        return result;
    }
};

File: 0201-0300/0227-Basic Calculator II/0227-Basic Calculator II.py
class Solution:
    def calculate(self, s: str) -> int:
        St = []
        num = 0
        preop = '+'
        for i in range(len(s)):
            if s[i].isdigit():
                num = num * 10 + ord(s[i]) - ord('0')
            if i == len(s) - 1 or s[i] in '+-*/':
                if preop == '+':
                    St.append(num)
                elif preop == '-':
                    St.append(-num)
                elif preop == '*':
                    St.append(St.pop() * num)
                else:
                    St.append(int(St.pop() / num))
                preop = s[i]
                num = 0
        return sum(St)
